### ISSUE TITLE
Support for X-SENDFILE on lighthttpd

### ISSUE THREAD

**Issue by dragotin** [owncloud/core#39547]
The maintainer of lighthttpd and asked why X-Sendfile support disabled and if it could be enabled again, for a win of performance:

```
Back in 2005, it looks like OwnCloud removed support for X-Sendfile, with a commit about locking. 
(https://github.com/owncloud/core/commit/8479702) 

I have taken a look through the NextCloud fork and made suggestions how to (selectively) enable X-Sendfile, and 
similar logic could probably be applied to OwnCloud, leading to an improvement in performance, as well as lower 
memory and lower CPU usage, when X-Sendfile is used. 
https://github.com/nextcloud/server/issues/29467 

I would be willing to mock up some patches to OwnCloud if there is interest and developers could give me guidance 
as to what is acceptable.
```
Maybe that should be reconsidered? @DeepDiver1975

**Comment by DeepDiver1975** [comment]
Guts feeling: we did remove x-sendfile support for a good reason and I cannot imagine a selective way of enabling it.

When ever there is a patch I happily review it and support the author with my knowledge

**Comment by gstrauss** [comment]
> Guts feeling: we did remove x-sendfile support for a good reason and I cannot imagine a selective way of enabling it.

Would you please help me to understand that reason?  What was the locking issue?  How does the locking issue manifest?

Does owncloud support partial PUT file upload?  If I have a directory containing large video files and I want to disable partial PUT file upload, is it now possible to enable X-Sendfile on this directory?  [Edit: based on your post about using sabre-dav, my hunch may be correct]

lighttpd mod_webdav supports file upload via PUT and atomically replaces files in the file system via upload to temp file and rename.  This can be inefficient for large files getting lots of small updates, but is not a common scenario for, say, a directory serving large videos.  If the performance of partial PUT is desired, then lighttpd has an option "deprecated-unsafe-partial-put" which edits the file in place, and though it is possible to corrupt downloads in progress in parallel to the modification, but possibly in the absence of using the webdav locking protocol.

Ultimately, I think X-Sendfile is possible and safe from a given directory if partial PUT is disabled for that directory, so I hope that you will take another look.  Thank you.

**Comment by gstrauss** [comment]
Multi-posting from https://github.com/nextcloud/server/issues/29467:

Thanks for the link.  Reading through those threads at owncloud/core#16997, my assessment remains that X-Sendfile is safe for use for a given directory (webdav collection) if partial PUT is disabled for that location.

Aside, partial PUT could be implemented by copying the original file to a temporary file, modifying the temporary file, and then atomically renaming the temporary file to the original name.  webdav locking would prevent multiple modifications at the same time. Alternatively, if the backing system supports version control, e.g. git, then X-Sendfile is safe when targetting a given revision, as modifications are made to a new revision.

In short, I believe that global support of partial PUT is extremely detrimental, imposing unnecessary requirements and performance overhead when downloading from directories containing large files which do not need to be edited in-place, and could instead use X-Sendfile for file downloads.

**Comment by DeepDiver1975** [comment]
It is not about partial PUT. 

ownCloud has a mechanism in place which protects files from being changed as an operation is working that file - we called it file locking.
Why is this needed:
Let's say a user is downloading a file - during that time this file is not allowed to be changed by any other operation. Once download finishes the file is unlocked and other operations can work on this file.
These other operations are e.g. overwrite the content, rename the file, move the file, move the parent folder and so on.
So we need to know when the download stops.

The idea to copy the original file and send this one out via x-sendfile might work - but only in limited cases. 

Finally as said above: sabre dav needs to support x-sendfiel first before oc or nc can profil :shrug:

**Comment by gstrauss** [comment]
For small files, partial PUT can be implemented with a full fill copy, edit, and atomic rename on the server.  A full file upload of a small file by a client, instead of partial PUT, is also reasonable.

For medium-sized and large files, a write heavy target (e.g. collaborative file editing) benefits from partial PUT, when the client supports partial PUT.

However, for medium-sized and large files which are seldom modified and frequently streamed -- such as a repository of large videos, music, and other media -- and for which there are seldom modifications once uploaded (write-once files), then supporting partial PUT is a complete waste if doing so precludes X-Sendfile streaming of those files.

Which webdav clients support (non-standardized) partial PUT?  If pages served from owncloud contain javascript implementing partial PUT, then could the page generation include a flag for whether or not the partial PUT code path in the javascript should be enabled or disabled for the target resource?

**Comment by gstrauss** [comment]
> Finally as said above: sabre dav needs to support x-sendfiel first before oc or nc can profil

The point I am trying to make is that if I disable partial PUT in sabre-dav, then what file operations are prevented in owncloud?  I feel that owncloud application-level oversight is needed, and that sabre-dav can not know whether X-Sendfile support is preferred for some scenarios and partial PUT support is desired for others.

If file modifications are performed atomically, as they are done in lighttpd mod_webdav, then the locking is handled properly.

Once I have an open file handle in the webserver, then even if the file is moved, the existing client download proceeds.  That said, there is a small race condition with X-Sendfile where a PHP backend sends X-Sendfile, but the front-end webserver has not yet opened the file pointed to by X-Sendfile, and another request comes in to modify the same file.  Again, this is typically a non-issue for large media directories which are effectively write-once and stream-many.

My argument remains that X-Sendfile would be very beneficial for many, but not all, directories, as long as partial PUT is not permitted for those locations.  For the locations requiring partial PUT, then X-Sendfile must not be used, and this knowledge is at the owncloud application level for whether or not X-Sendfile is configured to be enabled or not, or if owncloud is serving a location where partial PUT support is desired.

In short, I believe that partial PUT should be disabled by default and X-Sendfile enabled by default, and that locations serving apps which modify files in place, and for which the webdav client supports partial PUT, then only in those locations should X-Sendfile be disabled.

**Comment by gstrauss** [comment]
> It is not about partial PUT.

> ownCloud has a mechanism in place which protects files from being changed as an operation is working that file - we called it file locking.

I contend that partial PUT is the only webdav scenario which modifies the file in-place.  File renames, file moves, folder moves, and even file deletions do not affect an X-Sendfile download from the web server *once the web server has an open file handle to the X-Sendfile target*.

For write-once, stream-many media file locations, allowing X-Sendfile should be an easy win.  Configuring a location as write-once, stream-many is a configuration-level option.  Are you suggesting that such designations should be done in sabre-dav config rather than in owncloud config?  I think it should be in owncloud config.

**Comment by DeepDiver1975** [comment]
> I contend that partial PUT is the only webdav scenario which modifies the file in-place. File renames, file moves, folder moves, and even file deletions do not affect an X-Sendfile download from the web server _once the web server has an open file handle to the X-Sendfile target_.

this is not how owncloud works at this level. Here is the scenario:
there is a folder with multiple files and we want to move it to a different location (maybe from local disk to some mounted smb or what ever). oc will lock the folder and all it's children and then move each file one by one. 
If a download is in progress on one of these files the move operation is terminated even before one of them is moved - so to say as atomic operation on a higher level.
As we don't have an understanding when the download via x-send ends we cannot properly unlock the file.

Therefore I honestly don't see x-sendfile working in oc this way.
The idea of working on a copy might be of better approach - but no idea on how to implement that right away

finally: we are not using partial puts anyhow so you don't need to think about these implications ;-)

**Comment by gstrauss** [comment]
> Here is the scenario:
there is a folder with multiple files and we want to move it to a different location (maybe from local disk to some mounted smb or what ever). oc will lock the folder and all it's children and then move each file one by one.

This is not an issue once the webserver has an open file handle to the X-Sendfile target.  You can `rename` or `unlink` the file and the web server will still read from the file descriptor and send the contents to the client.  The kernel and the filesystem manage the rest.  If the file was `unlink`ed from the filesystem, the open file descriptor in the web server still has access to the file.  (Same thing if the file is `rename`d) Once all open file descriptors to the file are closed (i.e. file downloads are complete), the filesystem will be able to reclaim the blocks on disk used by the `unlink`ed file.  Neither the webserver nor owncloud need to worry about that.

**Comment by gstrauss** [comment]
I should also note that if webdav clients are well-behaved and fully implement webdav locking ("exclusive", "shared", "write"), then the proper implementation of webdav locking on the server will prevent other webdav operations when the locks are held.

Also, much of the above discussion about partial PUT and X-Sendfile is overly-cautious and assumes that clients may only implement "write" locking, or might not implement locking at all.

**Comment by gstrauss** [comment]
Similar to what I sketched out in https://github.com/nextcloud/server/issues/29467, I believe that files that are not encrypted and are local (not located on external storage) are eligible for X-Sendfile from owncloud.

**Comment by gstrauss** [comment]
@DeepDiver1975 I would like continue this conversation.  Can this issue please be re-opened for discussion?

Do you have any questions about the groundwork on which I am basing my suggestions?  Is there a peer you can loop in here to validate my statements about the WebDAV protocol and about POSIX-compliant local filesystem semantics?

**Comment by yevon** [comment]
Hi all! Are there any news about this? I was searching why did nextcloud / owncloud offer slow download speeds until I came across this. From what I could understand reading this, sendfile was deleted in the past due to file locking concerns. If you cannot know if the file was downloaded, you cannot unlock the file as it is read in background by the server apache /nginx process. And you don't want to make a temporary copy of the file due to size and resource consumption concerns. It would not be possible to just do a double verification to check if the file is locked? There should be a way to check if a file is being used by any other process isn't it? So you should check if the file is locked in the database, and if it is being used by any other process simultaneously. Wouldn't that solve the issue?
